modelforge/dataset/dataset.py

```diff
@@ -476,67 +508,265 @@ def __init__(
         split: SplittingStrategy = RandomRecordSplittingStrategy(),
         batch_size: int = 64,
         split_file: Optional[str] = None,
+        transform: Optional[nn.Module] = None,
     ):
         super().__init__()
         self.data = data
         self.batch_size = batch_size
-        self.split_idxs: Optional[str] = None
-        self.train_dataset = Optional[TorchDataset]
-        self.test_dataset = Optional[TorchDataset]
-        self.val_dataset = Optional[TorchDataset]
+        self.split_idxs = None
+        self.train_dataset = None
+        self.test_dataset = None
+        self.val_dataset = None
+        self.transform = transform
+        self.split = split
+        self.split_file = split_file
+        self.dataset_statistics = {}
+
+    @classmethod
+    def calculate_self_energies(
+        cls, dataset: TorchDataset, collate: bool = True
+    ) -> Dict[str, float]:
+        """
+        Calculates the self energies for each atomic number in the dataset by performing a least squares regression.
 
-        if split_file:
-            import numpy as np
+        Parameters
+        ----------
+        dataset : TorchDataset
+            The dataset from which to calculate self energies.
+        collate : bool, optional
+            If True, uses a custom collate function to gather batch data. Defaults to True.
 
-            logger.debug(f"Loading split indices from {split_file}")
-            self.split_idxs = np.load(split_file)
-        else:
-            logger.debug(f"Using splitting strategy {split}")
-            self.split = split
+        Returns
+        -------
+        Dict[int, float]
+            A dictionary mapping atomic numbers to their calculated self energies.
+        """
+        log.info("Computing self energies for elements in the dataset.")
+
+        # Define the collate function based on the `collate` parameter
+        collate_fn = collate_conformers if collate else None
+
+        # Initialize variables to hold data for regression
+        batch_size = 64
```

Comment:
I'm guessing this wouldn't be set directly by a user, but rather set explicitly when called by prepare data.  My thought for why this should be modifiable is from past experience having to modify codes because, e.g. character buffers were not set sufficiently high enough and there was no way to easily change it. More thinking that buffer sizes might be something that we may want to up as newer hardware comes out (or be determined on the fly based on hardware).   I just, in general dislike hardcoded variables deep in code; it's much easier to have those at the upper levels, even if not directly modifiable (as in, have these be set in the function that calls it).

Reply:
I understand. I will expose this in the upper levels and make it user modifiable.